### PR TITLE
Th2 1720

### PR DESCRIPTION
**Pull request by eugene-zheltov**

### CODE REVIEW

File: cradle-cassandra/src/main/java/com/exactpro/cradle/cassandra/CassandraCradleStorage.java [hunk 1]
```diff
@@ -336,18 +327,37 @@ protected void doStoreTestEventMessagesLink(StoredTestEventId eventId, StoredTes
 			throw new IOException("Error while storing link between test event "+eventId+" and "+messageIds.size()+" message(s)", e);
 		}
 	}
-	
+
 	@Override
 	protected CompletableFuture<Void> doStoreTestEventMessagesLinkAsync(StoredTestEventId eventId,
-			StoredTestEventId batchId, Collection<StoredMessageId> messageIds)
+																		StoredTestEventId batchId, Collection<StoredMessageId> messageIds)
```
Comment (Vladimir-Panarin-Exactpro) [inline cradle-cassandra/src/main/java/com/exactpro/cradle/cassandra/CassandraCradleStorage.java:333]: Pretty long shift for no reason.

File: cradle-cassandra/src/main/java/com/exactpro/cradle/cassandra/CassandraCradleStorage.java [hunk 2]
```diff
@@ -336,18 +327,37 @@ protected void doStoreTestEventMessagesLink(StoredTestEventId eventId, StoredTes
...
+
 	@Override
 	protected CompletableFuture<Void> doStoreTestEventMessagesLinkAsync(StoredTestEventId eventId,
-			StoredTestEventId batchId, Collection<StoredMessageId> messageIds)
+																		StoredTestEventId batchId, Collection<StoredMessageId> messageIds)
 	{
 		List<String> messageIdsStrings = messageIds.stream().map(StoredMessageId::toString).collect(toList());
 		String eventIdString = eventId.toString();
-		return CompletableFuture.allOf(storeMessagesOfTestEvent(eventIdString, messageIdsStrings),
-				storeTestEventOfMessages(messageIdsStrings, eventIdString, batchId));
```
Comment (Vladimir-Panarin-Exactpro) [inline cradle-cassandra/src/main/java/com/exactpro/cradle/cassandra/CassandraCradleStorage.java:347]: Removal of `storeTestEventOfMessages` is not part of this task, I think.

File: cradle-cassandra/src/main/java/com/exactpro/cradle/cassandra/CassandraCradleStorage.java [hunk 3]
```diff
@@ -412,28 +422,28 @@ protected StoredMessage doGetProcessedMessage(StoredMessageId id) throws IOExcep
...
 	protected long doGetLastProcessedMessageIndex(String streamName, Direction direction) throws IOException
 	{
 		return getLastIndex(ops.getProcessedMessageBatchOperator(), streamName, direction);
 	}
-	
+
 	@Override
 	protected StoredMessageId doGetNearestMessageId(String streamName, Direction direction, Instant timestamp,
-			TimeRelation timeRelation) throws IOException
+													TimeRelation timeRelation) throws IOException
```
Comment (Vladimir-Panarin-Exactpro) [inline cradle-cassandra/src/main/java/com/exactpro/cradle/cassandra/CassandraCradleStorage.java:446]: Long shift again.

File: cradle-cassandra/src/main/java/com/exactpro/cradle/cassandra/CassandraCradleStorage.java [hunk 4]
```diff
@@ -586,25 +596,25 @@ public TestEventsMessagesLinker getTestEventsMessagesLinker()
 			throw new IOException("Error while getting child test events", e);
 		}
 	}
-	
+
 	@Override
 	protected CompletableFuture<Iterable<StoredTestEventMetadata>> doGetTestEventsAsync(StoredTestEventId parentId,
-			Instant from, Instant to) throws CradleStorageException
+																						Instant from, Instant to) throws CradleStorageException
```
Comment (Vladimir-Panarin-Exactpro) [inline cradle-cassandra/src/main/java/com/exactpro/cradle/cassandra/CassandraCradleStorage.java:602]: Loooooong shift!

File: cradle-cassandra/src/main/java/com/exactpro/cradle/cassandra/CassandraStorageSettings.java [hunk 5]
```diff
@@ -54,7 +55,9 @@
 			testEventsChildrenTableName,
 			testEventsChildrenDatesTableName,
 			testEventsMessagesTableName,
-			messagesTestEventsTableName;
+			messagesTestEventsTableName,
+			recoveryStatesTableName,
```
Comment (Vladimir-Panarin-Exactpro) [inline cradle-cassandra/src/main/java/com/exactpro/cradle/cassandra/CassandraStorageSettings.java:59]: recoveryStatesTableName is not initialized. Is it needed?

File: cradle-cassandra/src/main/java/com/exactpro/cradle/cassandra/TablesCreator.java [hunk 6]
```diff
@@ -297,6 +298,23 @@ public void createMessagesTestEventsTable() throws IOException
...
+
+	public void createHealingIntervalsTable() throws IOException
+	{
+		String tableName = settings.getHealingIntervalsTableName();
+		if (isTableExists(tableName))
+			return;
+
+		CreateTable create = SchemaBuilder.createTable(settings.getKeyspace(), tableName).ifNotExists()
+				.withPartitionKey(INSTANCE_ID, DataTypes.UUID)
+				.withPartitionKey(HEALING_INTERVAL_ID, DataTypes.TEXT)
```
Comment (Vladimir-Panarin-Exactpro) [inline cradle-cassandra/src/main/java/com/exactpro/cradle/cassandra/TablesCreator.java:310]: With this table you can write interval data, but to get interval you'll need to know its ID. I doubt if crawlers actually know these IDs, they only know timestamp intervals to work in.

File: cradle-cassandra/src/main/java/com/exactpro/cradle/cassandra/dao/healing/HealingIntervalEntity.java [hunk 7]
```diff
@@ -0,0 +1,98 @@
+package com.exactpro.cradle.cassandra.dao.healing;
```
Comment (Vladimir-Panarin-Exactpro) [inline cradle-cassandra/src/main/java/com/exactpro/cradle/cassandra/dao/healing/HealingIntervalEntity.java:1]: Copyright!

File: cradle-cassandra/src/main/java/com/exactpro/cradle/cassandra/dao/healing/HealingIntervalEntity.java [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+    public HealingIntervalEntity(HealingInterval interval, UUID instanceId)
+    {
+        this.healingIntervalId = interval.getId();
+        this.startTime = interval.getStartTime();
+        this.endTime = interval.getEndTime();
+
+        try {
+            this.recoveryStateJson = mapper.writeValueAsString(interval.getRecoveryState());
+        } catch (JsonProcessingException e) {
+            logger.error("Error with converting recovery state to JSON format", e);
```
Comment (Vladimir-Panarin-Exactpro) [inline cradle-cassandra/src/main/java/com/exactpro/cradle/cassandra/dao/healing/HealingIntervalEntity.java:60]: `Error while converting...`

File: cradle-cassandra/src/main/java/com/exactpro/cradle/cassandra/dao/healing/HealingIntervalEntity.java [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+    public LocalTime getEndTime() { return endTime; }
+
+    public void setEndTime(LocalTime endTime) { this.endTime = endTime; }
+
+    public String getRecoveryStateJson() { return recoveryStateJson; }
+
+    public void setRecoveryStateJson(String recoveryStateJson) { this.recoveryStateJson = recoveryStateJson; }
+
+    public HealingInterval asHealingInterval() throws IOException {
+        logger.debug("RECOVERY STATE Json: " + recoveryStateJson);
```
Comment (Vladimir-Panarin-Exactpro) [inline cradle-cassandra/src/main/java/com/exactpro/cradle/cassandra/dao/healing/HealingIntervalEntity.java:93]: Debug output (if it is needed here) should use toString() to print info as one text. 
In current implementation output from other classes may appear between calls to logger.debug.

File: cradle-cassandra/src/main/java/com/exactpro/cradle/cassandra/dao/healing/HealingIntervalOperator.java [hunk 10]
```diff
@@ -0,0 +1,25 @@
+package com.exactpro.cradle.cassandra.dao.healing;
```
Comment (Vladimir-Panarin-Exactpro) [inline cradle-cassandra/src/main/java/com/exactpro/cradle/cassandra/dao/healing/HealingIntervalOperator.java:1]: Copyright!

File: cradle-core/src/main/java/com/exactpro/cradle/healing/HealingInterval.java [hunk 11]
```diff
@@ -0,0 +1,43 @@
+package com.exactpro.cradle.healing;
```
Comment (Vladimir-Panarin-Exactpro) [inline cradle-core/src/main/java/com/exactpro/cradle/healing/HealingInterval.java:1]: Copyright!

**Review by Vladimir-Panarin-Exactpro — CHANGES_REQUESTED** [review]
And better switch off auto-formatting. Now CassadnraCradleStorage is full of "changes" related to tab/space conversion and code shift adjustments.

File: cradle-cassandra/src/main/java/com/exactpro/cradle/cassandra/CassandraCradleStorage.java (hunk 2 above)
Comment (eugene-zheltov) [reply cradle-cassandra/src/main/java/com/exactpro/cradle/cassandra/CassandraCradleStorage.java:347]: It was in with @Nikita-Smirnov-Exactpro in order to implement the linking of messages with their events. I split this method into two methods (added doStoreMessageTestEventLinkAsync(eventId, batchId, messageIds)).

File: cradle-cassandra/src/main/java/com/exactpro/cradle/cassandra/CassandraCradleStorage.java [hunk 12]
```diff
@@ -109,19 +103,19 @@
 	private final CassandraStorageSettings settings;
 	private final CassandraSemaphore semaphore;
 	private final CradleObjectsFactory objectsFactory;
-	
+
```
Comment (Vladimir-Panarin-Exactpro) [inline cradle-cassandra/src/main/java/com/exactpro/cradle/cassandra/CassandraCradleStorage.java:106]: This ckass is full of "changes" due to presence of stuff like this, i.e. empty line or tabs removal. Please disable this feature in your IDE, if you don't mind.

**Review by Vladimir-Panarin-Exactpro — CHANGES_REQUESTED** [review]

File: cradle-cassandra/src/main/java/com/exactpro/cradle/cassandra/dao/intervals/IntervalEntity.java [hunk 13]
```diff
@@ -0,0 +1,129 @@
+package com.exactpro.cradle.cassandra.dao.intervals;
```
Comment (Vladimir-Panarin-Exactpro) [inline cradle-cassandra/src/main/java/com/exactpro/cradle/cassandra/dao/intervals/IntervalEntity.java:1]: Add copyright.

File: cradle-cassandra/src/main/java/com/exactpro/cradle/cassandra/dao/intervals/IntervalOperator.java [hunk 14]
```diff
@@ -0,0 +1,32 @@
+package com.exactpro.cradle.cassandra.dao.intervals;
```
Comment (Vladimir-Panarin-Exactpro) [inline cradle-cassandra/src/main/java/com/exactpro/cradle/cassandra/dao/intervals/IntervalOperator.java:1]: Add copyright.

File: cradle-core/src/main/java/com/exactpro/cradle/CradleStorage.java [hunk 15]
```diff
@@ -289,41 +308,73 @@ public final void storeTestEvent(StoredTestEvent event) throws IOException
...
 	 * @param messagesIds collection of stored message IDs
 	 * @throws IOException if data writing failed
 	 */
 	public final void storeTestEventMessagesLink(StoredTestEventId eventId, StoredTestEventId batchId, Collection<StoredMessageId> messagesIds) throws IOException
 	{
-		logger.debug("Storing links between test event {} and {} message(s)", eventId, messagesIds.size());
+		logger.debug("Storing {} message(s) of test event {}", messagesIds.size(), eventId);
 		doStoreTestEventMessagesLink(eventId, batchId, messagesIds);
-		logger.debug("Links between test event {} and {} message(s) have been stored", eventId, messagesIds.size());
+		logger.debug("Message(s) {} of test event {} have been stored", messagesIds.size(), eventId);
```
[the remark this answers is not in the document]
Comment (Vladimir-Panarin-Exactpro) [reply cradle-core/src/main/java/com/exactpro/cradle/CradleStorage.java:324]: UP.

File: cradle-core/src/main/java/com/exactpro/cradle/CradleStorage.java [hunk 16]
```diff
@@ -293,41 +310,73 @@ public final void storeTestEvent(StoredTestEvent event) throws IOException
...
-		logger.debug("Storing links between test event {} and {} message(s) asynchronously", eventId, messagesIds.size());
+		logger.debug("Storing {} message(s) of test event {}", messagesIds.size(), eventId);
 		return doStoreTestEventMessagesLinkAsync(eventId, batchId, messagesIds)
 				.whenComplete((r, error) -> {
 					if (error != null)
-						logger.error("Error while storing links between test event "+eventId+" and "+messagesIds.size()+" message(s) asynchronously", error);
+						logger.error("Error while storing "+messagesIds.size()+" message(s) of test event "+eventId+" asynchronously", error);
 					else
-						logger.debug("Links between test event {} and {} message(s) have been stored asynchronously", eventId, messagesIds.size());
+						logger.debug("Message(s) {} of test event {} have been stored asynchronously", messagesIds.size(), eventId);
```
Comment (Vladimir-Panarin-Exactpro) [inline cradle-core/src/main/java/com/exactpro/cradle/CradleStorage.java:342]: This method now has incorrect logging and Java doc. Former was correct.

File: cradle-core/src/main/java/com/exactpro/cradle/CradleStorage.java [hunk 17]
```diff
@@ -293,41 +310,73 @@ public final void storeTestEvent(StoredTestEvent event) throws IOException
...
+		logger.debug("Storing test event {} of {} message(s) asynchronously", eventId, messagesIds.size());
+		return doStoreMessageTestEventLinkAsync(eventId, batchId, messagesIds)
+				.whenComplete((r, error) -> {
+					if (error != null)
+						logger.error("Error while storing test event "+eventId+" of "+messagesIds.size()+" message(s) asynchronously", error);
+					else
+						logger.debug("Test event {} of {} message(s) has been stored asynchronously", eventId, messagesIds.size());
+				});
+
+	}
```
Comment (Vladimir-Panarin-Exactpro) [inline cradle-core/src/main/java/com/exactpro/cradle/CradleStorage.java:379]: Are these methods really needed? This is part of Cassandra-dependent implementation.

File: cradle-core/src/main/java/com/exactpro/cradle/CradleStorage.java [hunk 18]
```diff
@@ -849,4 +898,148 @@ protected void storeTimeMessages(Collection<StoredMessage> messages) throws IOEx
...
+		logger.debug("Updating interval with {}", interval.getId());
+		boolean result = this.doSetIntervalLastUpdateTimeAndDate(interval, newLastUpdateTime);
+
+		if (result) {
+			return true;
+		}
+		else {
+			logger.debug("Updated interval {}", interval.getId());
+			return false;
+		}
```
Comment (Vladimir-Panarin-Exactpro) [inline cradle-core/src/main/java/com/exactpro/cradle/CradleStorage.java:983]: Replace with this:
```if (!this.doSetIntervalLastUpdateTimeAndDate(interval, newLastUpdateTime)) {
logger.debug("Updated interval {}", interval.getId());
return false;
}
return true;
```

File: cradle-core/src/main/java/com/exactpro/cradle/CradleStorage.java [hunk 19]
```diff
@@ -849,4 +898,148 @@ protected void storeTimeMessages(Collection<StoredMessage> messages) throws IOEx
...
+	 */
+	public final CompletableFuture<Boolean> setIntevalLastUpdateTimeAndDateAsync(Interval interval, Instant newLastUpdateTime)
+	{
+		logger.debug("Asynchronously updating interval {}", interval.getId());
+		CompletableFuture<Boolean> result = doSetIntervalLastUpdateTimeAndDateAsync(interval, newLastUpdateTime)
+				.whenComplete((r, error) -> {
+					if (error != null)
+						logger.error("Error while asynchronously updating interval "+interval.getId());
+					else
+						logger.debug("Interval {} updated asynchronously", interval.getId());
```
Comment (Vladimir-Panarin-Exactpro) [inline cradle-core/src/main/java/com/exactpro/cradle/CradleStorage.java:1001]: Shouldn't you check r for true like in the method above?

File: cradle-core/src/main/java/com/exactpro/cradle/intervals/Interval.java [hunk 20]
```diff
@@ -0,0 +1,66 @@
+package com.exactpro.cradle.intervals;
```
Comment (Vladimir-Panarin-Exactpro) [inline cradle-core/src/main/java/com/exactpro/cradle/intervals/Interval.java:1]: Add copyright.

**Review by Vladimir-Panarin-Exactpro — CHANGES_REQUESTED** [review]

File: cradle-core/src/main/java/com/exactpro/cradle/CradleStorage.java [hunk 21]
```diff
@@ -89,7 +96,19 @@
...
+	protected abstract void doUpdateEventStatus(StoredTestEventWrapper event, boolean success) throws IOException;
+	protected abstract CompletableFuture<Void> doUpdateEventStatusAsync(StoredTestEventWrapper event, boolean success);
+	protected abstract void doUpdateRecoveryState(RecoveryState recoveryState, LocalDate healingIntervalStartDate, LocalTime healingIntervalStartTime, String crawlerType) throws IOException;
+	protected abstract CompletableFuture<Void> doUpdateRecoveryStateAsync(RecoveryState recoveryState, LocalDate healingIntervalStartDate, LocalTime healingIntervalStartTime, String crawlerType);
+	protected abstract void doStoreHealedEventsIds(Set<String> healedEventIds, LocalDate healingIntervalStartDate, LocalTime healingIntervalStartTime, String crawlerType) throws IOException;
+	protected abstract CompletableFuture<Void> doStoreHealedEventsIdsAsync(Set<String> healedEventIds, LocalDate healingIntervalStartDate, LocalTime healingIntervalStartTime, String crawlerType);
```
[the remark this answers is not in the document]
Comment (Vladimir-Panarin-Exactpro) [reply cradle-core/src/main/java/com/exactpro/cradle/CradleStorage.java]: UP.

**Review by Vladimir-Panarin-Exactpro — CHANGES_REQUESTED** [review]
Please extract methods to work with intervals into a separate class. This will make `CradleStorage` smaller and allow to focus on work with intervals by using separate object of that new class.
Other changes are ok.

File: cradle-cassandra/src/main/java/com/exactpro/cradle/cassandra/dao/intervals/CassandraIntervalsWorker.java [hunk 22]
```diff
@@ -0,0 +1,162 @@
...
+    @Override
+    public Iterable<Interval> getIntervals(Instant from, Instant to, String crawlerName, String crawlerVersion) throws IOException
+    {
+        try
+        {
+            return getIntervalsAsync(from, to, crawlerName, crawlerVersion).get();
+        }
+        catch (Exception e)
+        {
+            throw new IOException("Error while getting healing intervals from " + from, e);
```
Comment (Vladimir-Panarin-Exactpro) [inline cradle-cassandra/src/main/java/com/exactpro/cradle/cassandra/dao/intervals/CassandraIntervalsWorker.java:77]: They are not healing intervals, they are just intervals.
And you're getting them "from" and "to". `crawlerName` would also be useful in the error.

File: cradle-cassandra/src/main/java/com/exactpro/cradle/cassandra/dao/intervals/CassandraIntervalsWorker.java [hunk 23]
```diff
@@ -0,0 +1,162 @@
...
+                                .getIntervals(instanceUuid, date, fromTime, toTime, crawlerName, crawlerVersion, readAttrs));
+
+        return future.thenApply(entities -> {
+            try
+            {
+                return new IntervalsIteratorAdapter(entities);
+            }
+            catch (Exception error)
+            {
+                throw new CompletionException("Could not get healing intervals from "+from, error);
```
Comment (Vladimir-Panarin-Exactpro) [inline cradle-cassandra/src/main/java/com/exactpro/cradle/cassandra/dao/intervals/CassandraIntervalsWorker.java:104]: They are not healing intervals, they are just intervals.
And you're getting them "from" and "to". `crawlerName` would also be useful in the error.

File: cradle-cassandra/src/main/java/com/exactpro/cradle/cassandra/dao/intervals/CassandraIntervalsWorker.java [hunk 24]
```diff
@@ -0,0 +1,162 @@
+package com.exactpro.cradle.cassandra.dao.intervals;
```
Comment (Vladimir-Panarin-Exactpro) [inline cradle-cassandra/src/main/java/com/exactpro/cradle/cassandra/dao/intervals/CassandraIntervalsWorker.java:1]: Add copyright.

File: cradle-core/src/main/java/com/exactpro/cradle/intervals/IntervalsWorker.java [hunk 25]
```diff
@@ -0,0 +1,58 @@
+package com.exactpro.cradle.intervals;
```
Comment (Vladimir-Panarin-Exactpro) [inline cradle-core/src/main/java/com/exactpro/cradle/intervals/IntervalsWorker.java:1]: Add copyright.

**Review by Vladimir-Panarin-Exactpro — CHANGES_REQUESTED** [review]

**Review by Vladimir-Panarin-Exactpro — APPROVED** [review]
I haven't checked how this works, but code is ok.

File: cradle-cassandra/src/main/java/com/exactpro/cradle/cassandra/dao/intervals/CassandraIntervalsWorker.java [hunk 26]
```diff
@@ -171,9 +171,33 @@ public void updateRecoveryState(Interval interval, RecoveryState recoveryState)
...
+
+        return future.thenAccept(r -> {});
+    }
+
+    @Override
+    public void setIntervalProcessed(Interval interval, boolean processed) throws IOException
+    {
+        try
+        {
+            setIntervalProcessedAsync(interval, processed);
```
Comment (Vladimir-Panarin-Exactpro) [inline cradle-cassandra/src/main/java/com/exactpro/cradle/cassandra/dao/intervals/CassandraIntervalsWorker.java:186]: .get() is missing?

File: cradle-cassandra/src/main/java/com/exactpro/cradle/cassandra/dao/intervals/IntervalEntity.java [hunk 27]
```diff
@@ -45,8 +45,11 @@
     private String id;
 
     @ClusteringColumn(1)
```
Comment (Vladimir-Panarin-Exactpro) [inline cradle-cassandra/src/main/java/com/exactpro/cradle/cassandra/dao/intervals/IntervalEntity.java:47]: Better store all partition keys and clustering keys in the same order as in table definition.
Interval start date is not clustering column 1, it is 2.

File: cradle-core/src/main/java/com/exactpro/cradle/intervals/Interval.java [hunk 28]
```diff
@@ -63,19 +69,22 @@ public Interval(String id, LocalTime startTime, LocalTime endTime, LocalDate dat
...
                 .append("id=").append(id).append(",").append(CompressionUtils.EOL)
-                .append("date=").append(date).append(",").append(CompressionUtils.EOL)
+                .append("date=").append(startDate).append(",").append(CompressionUtils.EOL)
                 .append("startTime=").append(startTime).append(",").append(CompressionUtils.EOL)
                 .append("endTime=").append(endTime).append(",").append(CompressionUtils.EOL)
                 .append("lastUpdateDate=").append(lastUpdateDate).append(",").append(CompressionUtils.EOL)
                 .append("lastUpdateTime=").append(lastUpdateTime).append(",").append(CompressionUtils.EOL)
                 .append("crawlerName=").append(crawlerName).append(",").append(CompressionUtils.EOL)
                 .append("crawlerVersion=").append(crawlerVersion).append(",").append(CompressionUtils.EOL)
+                .append("processed=").append(processed).append(",").append(CompressionUtils.EOL)
```
Comment (Vladimir-Panarin-Exactpro) [inline cradle-core/src/main/java/com/exactpro/cradle/intervals/Interval.java:87]: End date is missing.

File: cradle-core/src/main/java/com/exactpro/cradle/intervals/Interval.java [hunk 29]
```diff
@@ -24,25 +24,29 @@
...
     private final LocalDate lastUpdateDate;
     private final LocalTime lastUpdateTime;
     private final String crawlerName;
     private final String crawlerVersion;
+    private final boolean processed;
 
-    public Interval(String id, LocalTime startTime, LocalTime endTime, LocalDate date, RecoveryState recoveryState,
-                    LocalDate lastUpdateDate, LocalTime lastUpdateTime, String crawlerName, String crawlerVersion)
+    public Interval(String id, LocalTime startTime, LocalTime endTime, LocalDate startDate, LocalDate endDate, RecoveryState recoveryState,
+                    LocalDate lastUpdateDate, LocalTime lastUpdateTime, String crawlerName, String crawlerVersion, boolean processed)
```
Comment (Vladimir-Panarin-Exactpro) [inline cradle-core/src/main/java/com/exactpro/cradle/intervals/Interval.java:37]: Too many parameters in constructor. Use builder instead?

**Review by Vladimir-Panarin-Exactpro — CHANGES_REQUESTED** [review]
In my opinion it would be more correct to partition intervals by date. Currently all of them are stored in one partition.
Partitioning them by date you'll be able to get all intervals of a day, ordered by time. Crawler name and version will be additional data to get or operate.

**Review by Vladimir-Panarin-Exactpro — APPROVED** [review]

**Review by Vladimir-Panarin-Exactpro — APPROVED** [review]

File: cradle-core/src/main/java/com/exactpro/cradle/intervals/IntervalsWorker.java [hunk 30]
```diff
@@ -38,18 +40,38 @@
...
+     * @param crawlerType type of Crawler
      * @return iterable of intervals
      */
-    Iterable<Interval> getIntervals(Instant from, Instant to, String crawlerName, String crawlerVersion, String crawlerType) throws IOException;
+    Iterable<Interval> getIntervalsPerDay(Instant from, Instant to, String crawlerName, String crawlerVersion, String crawlerType) throws IOException;
 
     /**
-     * Asynchronously obtains iterable of intervals with startTime greater or equal that from and less or equal then to
-     * @param from time from which intervals are being searched
+     * Asynchronously obtains iterable of intervals with startTime greater or equal that from and less or equal then to. Intervals must be within one day
```
Comment (Vladimir-Panarin-Exactpro) [inline cradle-core/src/main/java/com/exactpro/cradle/intervals/IntervalsWorker.java:54]: Replace this line this the following:
`* Asynchronously obtains iterable of intervals with startTime greater than or equal to "from" and less than or equal to "to". Intervals must be within one day`

File: cradle-cassandra/src/main/java/com/exactpro/cradle/cassandra/dao/intervals/CassandraIntervalsWorker.java [hunk 31]
```diff
@@ -125,6 +131,38 @@ public boolean storeInterval(Interval interval) throws IOException
...
 
+    @Override
+    public Iterable<Interval> getIntervals(Instant from, Instant to, String crawlerName, String crawlerVersion, String crawlerType) throws IOException
+    {
+        LocalDateTime fromDateTime = LocalDateTime.ofInstant(from, TIMEZONE_OFFSET),
+                toDateTime = LocalDateTime.ofInstant(to, TIMEZONE_OFFSET);
+
+        Iterable<Interval> result = new ArrayList<>();
+
+        LocalDateTime point = fromDateTime;
```
Comment (Vladimir-Panarin-Exactpro) [inline cradle-cassandra/src/main/java/com/exactpro/cradle/cassandra/dao/intervals/CassandraIntervalsWorker.java:142]: Move down to be before `while`.

File: cradle-cassandra/src/main/java/com/exactpro/cradle/cassandra/dao/intervals/CassandraIntervalsWorker.java [hunk 32]
```diff
@@ -125,6 +131,38 @@ public boolean storeInterval(Interval interval) throws IOException
...
+        if (fromDateTime.toLocalDate().compareTo(toDateTime.toLocalDate()) == 0)
+        {
+            return getIntervalsPerDay(from, to, crawlerName, crawlerVersion, crawlerType);
+        }
+
+        while (point.isBefore(toDateTime))
+        {
+            point = LocalDateTime.of(fromDateTime.toLocalDate(), LocalTime.MAX);
+
+            if (point.isAfter(toDateTime) || point.isEqual(toDateTime))
```
Comment (Vladimir-Panarin-Exactpro) [inline cradle-cassandra/src/main/java/com/exactpro/cradle/cassandra/dao/intervals/CassandraIntervalsWorker.java:153]: Looks like `point.isEqual(toDateTime)` is meaningless.

**Pull request merged**